File: 백준/Platinum/1725. 히스토그램/히스토그램.cc
#include<iostream>
#include<stack>
#include<vector>

using namespace std;

int main() {

    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    int N, ans = 0;
    stack<int> stack;

    //입력
    cin >> N;
    vector<int> h(N + 2);   //높이
    for (int i = 1; i <= N; i++) {
        cin >> h[i];
    }

    //연산
    stack.push(0);
    for (int i = 1; i <= N + 1; i++) {
        //현재 높이가 이전 높이보다 낮을 때
        while (!stack.empty() && h[stack.top()] > h[i]) {
            int height = h[stack.top()];
            stack.pop();
            int width = i - stack.top() - 1;
            ans = max(ans, height * width);
        }
        stack.push(i);
    }

    //출력
    cout << ans;

    return 0;
}

File: 백준/Platinum/1725. 히스토그램/히스토그램.java
import java.util.*;
import java.io.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());

		long[] h = new long[n + 2];
		for (int i = 1; i <= n; i++) {
			h[i] = Long.parseLong(br.readLine());
		} // 입력 끝

		Stack<Integer> stack = new Stack<>();
		long ans = 0;
		stack.push(0);
		for (int i = 1; i <= n + 1; i++) {
			while (!stack.isEmpty() && h[i] < h[stack.peek()]) {
				ans = Math.max(ans, h[stack.pop()] * (i - stack.peek() - 1));
			}
			stack.push(i);
		}
		System.out.println(ans);
	}
}
